### ISSUE TITLE
Development Roadmap (2025 Q3)

### ISSUE THREAD

**Issue by zhyncs** [sgl-project/sglang#7736]
Here is the development roadmap for 2025 H2. Contributions and feedback are welcome ([Join Slack](https://slack.sglang.ai)). The previous 2025 H1 roadmap can be found in https://github.com/sgl-project/sglang/issues/4042

## Focus
- Feature compatibility and reliability: Make all advanced features fully compatible with each other and achieve production-level reliability, such as P/D disaggregation, all parallelisms, speculative decoding, and load balancing.
- Usability: easy installation on all backends; simple launch scripts for large-scale deployments.
- Kernel optimizations for new generations of hardware (Blackwell, MI350, TPU, etc).
- Reinforcement learning training framework integration.

## Core refactor
- Simplify overlap scheduler #11762
- Piecewise cuda graph and torch compile support  #11490
- Flexible memory cache layer (mem_cache_v2) #11313
- Document major parts
- Remove deadcode (e.g, double sparsity, flashinfer lora backend)

## Speculative decoding
- overlap scheduler #11762
- Reference-based speculative decoding #9873 
- Make speculative decoding compatible with all other features 
- Fix all corner cases in structured output + speculative decoding + reasoning/function call parsing

## KVCache system
- HiCache with storage layer enhancement #7704
  - Mooncake integration: https://github.com/sgl-project/sglang/pull/7211
  - 3FS integration: https://github.com/sgl-project/sglang/pull/7280
  - More backends including https://github.com/ai-dynamo/nixl, etc.
  - Alternatively, directly mapping storage device or remote memory pool as host memory pool if hardware permits: https://github.com/sgl-project/sglang/pull/7896
- Holistic scheduling with a global view of KV cache, more can be found: #8210
- C++ implementation of radix and hiradix tree to mitigate scheduling overhead #7194 #7369 #6907
- Sliding window attention memory pool optimizations (dynamic ratio)

## Kernel
- FP4 moe kernels #7711
- More attention kernel variants from flashinfer #7938
- CuDNN attention backend #5505 #7841
- Torch FlexAttention backend #9947 
- Overlap allreduce in tensor parallelism #9058 
- More allreduce optimizations (e.g., multi-node nvlink, fusion)  #7621
- Improve FP8 gemm kernels on hopper and blackwell

## Parallelism
- Support arbitrary combination of DP attention + TP + EP #7966
- Async schedule for pipeline parallelism #11857 
- Elastic EP #11837
- Support all parallelism + speculative decoding
- Support all parallelism + PD disaggregation

## PD Disaggregation
- CPU transfer backend #8210
- Auto scaling in OME 
- Fault tolerance & reliability & health check

## Quantization
- Major roadmap: https://github.com/sgl-project/sglang/issues/8180
- ModelOpt integration to support all models
- Communication quantization (fp8 allreduce)
- MXFP4 support 

## RL framework integration 
- AREAL, slime, veRL integration (sorted alphabetically) 
- Faster weight sync
- Reproduce Deepseek/Kimi + GRPO training

## Multi-LoRA serving
- Major Roadmap: #2929
- Asynchronous lora updates: #8213 
- Support Radix Attention: #7216 
- GPU pinning: #8697 
- LRU cache: #11041
- Performance improvements #8940 
- Support embedding layers #8222 
- Serving lora adaptors with different ranks using unified paging 

## Hardware
- Better CI coverage of AMD
- Support Ascend NPU #8004
- H20-specific optimizations #8151
- Support TPU https://github.com/sgl-project/sglang-jax
- Support intel XPU
- Support more hardware with multi-framework #8199

## Model coverage
- Day 0 support for all upcoming OSS models
    - GPT OSS #8833
    - DeepSeek v3.2 #11060 #11989
- Multi-modal models #
- Language models
  -  [ ] Hybrid Mamba/linear attention models #10233 #10909

## API layer
- Provide an gRPC interface layer #10341 #11986
- Rewrite api layer (fastapi, tokenizer manager) in sgl-router  #10341 #11986
- Support all advanced apis (e.g. OpenAI response API, MCP integration) #10341 #11986

**Comment by mickqian** [comment]
Should we include [encode-prefill-decode disaggregation](https://arxiv.org/abs/2502.00937)?  ref https://github.com/sgl-project/sglang/issues/8223

**Comment by zzt93** [comment]
@zhyncs "Async schedule for pipeline parallelism" I am developing this feature, can you assign it to me?

<img width="2506" height="1008" alt="Image" src="https://github.com/user-attachments/assets/8eb7cb32-bf77-42a5-81a7-5108bb010675" />

**Comment by QiuMike** [comment]
> Should we include [encode-prefill-decode disaggregation](https://arxiv.org/abs/2502.00937)? ref [#8223](https://github.com/sgl-project/sglang/issues/8223)
@mickqian PR https://github.com/sgl-project/sglang/pull/7933 has already implemented EPD, and shows good performance improvement, any reason to refile another issue to implement another one?

**Comment by zhyncs** [comment]
Hi @zzt93 Thanks! We have finished the development, and expect to upstream in the next two days.

**Comment by ZelinMa557** [comment]
Interested in adding support for mamba/hybrid models

**Comment by johnnynunez** [comment]
@zhyncs democratize aarch64 wheels distribution in pypi for gh200,gb200/300 and thor/spark

**Comment by yukiy927** [comment]
I might be misunderstanding, but “FlexPrefill” (paper-defined sparse attention, often FA3-based) and PyTorch “FlexAttention” are distinct. In PR [#6867](https://github.com/sgl-project/sglang/pull/6867) the title reads “Torch FlexAttention backend: add flexprefill…”, which could be confusing. Does this implement the FlexPrefill pattern via Torch FlexAttention, or FA3 kernels, or both? A brief note/rename could help clarify. Thanks!

> * Torch FlexAttention backend [[Kernel] feat: add flexprefill for long-context #6867](https://github.com/sgl-project/sglang/pull/6867)

**Comment by JoursBleu** [comment]
It seems that PR #9947 is the correct implementation of the Torch FlexAttention backend.


> I might be misunderstanding, but “FlexPrefill” (paper-defined sparse attention, often FA3-based) and PyTorch “FlexAttention” are distinct. In PR [#6867](https://github.com/sgl-project/sglang/pull/6867) the title reads “Torch FlexAttention backend: add flexprefill…”, which could be confusing. Does this implement the FlexPrefill pattern via Torch FlexAttention, or FA3 kernels, or both? A brief note/rename could help clarify. Thanks!
> 
> > * Torch FlexAttention backend [[Kernel] feat: add flexprefill for long-context #6867](https://github.com/sgl-project/sglang/pull/6867)

**Comment by vincentzed** [comment]
Will want to support the decode context parallel [here](https://arxiv.org/html/2411.01783v2). And use it with flashmla or FI.

**Comment by HaibinLai** [comment]
Really exciting to see all parallelism modes planned for speculative decoding!
Our current workload relies heavily on pipeline parallelism (PP), so support for speculative decoding + PP would be extremely valuable for us.

I was wondering — will there be a more detailed design roadmap or pr tracking this work (especially for PP integration)? I’d love to follow or contribute. Thanks for pushing this forward!

**Comment by merrymercy** [comment]
@HaibinLai  see the new roadmap https://github.com/sgl-project/sglang/issues/12780 and https://github.com/sgl-project/sglang/issues/11857
